core/src/price_estimation/average_price_source.rs

```diff
@@ -54,7 +54,7 @@ fn lossless_merge<T: Eq + Hash, U>(map_collection: Vec<HashMap<T, U>>) -> HashMa
     gathered_maps
 }
 
-fn average_prices(price_maps: Vec<HashMap<TokenId, u128>>) -> HashMap<TokenId, u128> {
+pub fn average_prices(price_maps: Vec<HashMap<TokenId, u128>>) -> HashMap<TokenId, u128> {
```

Comment:
We use this function directly because we do not want to store just an AveragePriceSource because we update the pricegraph based price source manually.

Reply:
Maybe this means we need to refactor the `AveragePriceSource` abstraction a bit? Alternatively, maybe we need to implement an `UpdateingOrderbookBased` price source in `price-estimator` that can integrate with `AveragePriceSource` but be updated by external triggers?

Reply:
I had the same thought about AveragePriceSource but the abstractions I came up with were all overly complex for what I want to accomplish here.
Your second idea is simpler but it still have the same feeling that the amount of extra code needed is not worth it.
To me there is no problem having this function public. This is the most generic form of calculating the average because it does not care about how the prices are fetched or who owns the price sources.

Reply:
> takes ownership of the inner price sources

One idea is we can:
```
impl<T: PriceSource> for Arc<T> { ... }
```

Then we can `Box<Arc<T>>` to share the inner price source? Alternatively, we can just make the `AveragePriceSource` take `Arc`.

Reply:
Don't we already have a ThreadedPriceSource, which is basically a cache that can get updated externally (by a timeout running in a separate thread). Maybe we can make that more generic that it can also be updated by a synchronous trigger.